spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -368,6 +374,23 @@ private void commitWithSnapshotIsolation(OverwriteFiles overwriteFiles,
     }
   }
 
+  private class CompactionWrite extends BaseBatchWrite {
+    private final String jobID;
+
+    private CompactionWrite(String jobID) {
+      this.jobID = jobID;
+    }
+
+    @Override
+    public void commit(WriterCommitMessage[] messages) {
+      List<DataFile> newDataFiles = Lists.newArrayListWithExpectedSize(messages.length);
+      for (DataFile file : files(messages)) {
+        newDataFiles.add(file);
+      }
+      DataCompactionJobCoordinator.commitJob(table, jobID, newDataFiles);
```

Comment:
The commit just propagates new files to the job coordinator.

Reply:
Why not do the actual commit to the table here and fetch the files to replace? I think it would be much cleaner if the commit happened as part of the write. Then a failure to commit would use normal cleanup and would show up as an error in Spark, rather than having the controlling code do the commit.

Reply:
Internally, we are doing actual commits during writes. However, we want to process different partitions using separate jobs. For now, we simply generate a number of commits for a single round of compaction. How can we handle this place if we should be able to commit everything as one operation if the user asks for that?

In our implementation, we do synchronization on the driver to make sure only one job is trying to commit even though we submit and run a number of partition jobs concurrently. No need to generate extra conflicts.

Reply:
cc @RussellSpitzer for thoughts

Reply:
I think the commit should be done by the caller as well. Just because this let's us do more complicated behaviors with the files written like the grouping/splitting of commits independent of Spark write/shuffle jobs.

Reply:
If we are to make it more generic and call like `TaskSetManager` (which is a good thing), we should probably remove handling the job result from this class. Instead, we will need a data compaction specific class that would allow us to do locking from the same driver between commits from multiple jobs. We can make it handle the results too.

Reply:
I am thinking about something along these lines:

```
// manages task sets
TaskSetManager.addTaskSet(table, id, tasks)
TaskSetManager.removeTaskSet(table, id)

// coordinates compaction jobs and does single JVM locking
DataCompactionJobCoordinator.finishJob(table, id, newDataFiles)
DataCompactionJobCoordinator.commitJob(table, id)
DataCompactionJobCoordinator.commitJobs(table, ids)
```

Having the commit logic in `DataCompactionJobCoordinator` should simplify the action itself.

Reply:
Thoughts, @RussellSpitzer @rdblue?

Reply:
DCJC would still need to return the files or the compaction action wouldn't know what actually changed, which I think we want to still expose? I also feel like we want to have logging kind of like I put in the prototype which exposes where in a general Action you are which requires knowing which Id's are part of a particular action in the DCJC?

Reply:
I think the action can do whatever it does in the WIP PR now (i.e. plan jobs, log partitions, num files, etc), then add task sets into `TaskSetManager`, trigger compaction jobs with read/write options, then the writer will call `finishJob` to propagate (but not commit) the results and then the action will call `commitJob`/`commitJobs`.

We will only have to return correct stats from `commitJob` (basically, whatever is needed by the result). We can make the job coordinator part of the action too.

The logic in the action won't be trivial so I am thinking whether we can simplify it by moving parts into helper objects. In this case, the coordinator can take of committing and locking.

Reply:
The task manager part alone is in PR #2472.